rustls/tests/api.rs

```diff
@@ -5134,6 +5133,68 @@ fn test_server_picks_dhe_group_when_clienthello_has_no_dhe_group_in_named_groups
     assert!(server.process_new_packets().is_ok());
 }
 
+#[cfg(feature = "tls12")]
+#[test]
+fn test_server_avoids_dhe_cipher_suites_when_client_has_no_known_dhe_in_groups_ext() {
+    use rustls::{crypto::SupportedKxGroup, NamedGroup};
+
+    #[derive(Debug)]
+    struct FakeFFDHE4096Kx;
+    impl SupportedKxGroup for FakeFFDHE4096Kx {
+        fn start(&self) -> Result<Box<dyn rustls::crypto::ActiveKeyExchange>, Error> {
+            unimplemented!()
+        }
+        fn name(&self) -> NamedGroup {
+            NamedGroup::FFDHE4096
+        }
+    }
+
+    let client_config = finish_client_config(
+        KeyType::Rsa,
+        rustls::ClientConfig::builder_with_provider(
+            CryptoProvider {
+                cipher_suites: vec![
+                    ffdhe::TLS_DHE_RSA_WITH_AES_128_GCM_SHA256,
+                    provider::cipher_suite::TLS_ECDHE_RSA_WITH_AES_128_GCM_SHA256,
+                ],
+                kx_groups: vec![&FakeFFDHE4096Kx, provider::kx_group::SECP256R1],
+                ..provider::default_provider()
+            }
+            .into(),
+        )
+        .with_safe_default_protocol_versions()
+        .unwrap(),
+    );
+
+    let server_config = finish_server_config(
+        KeyType::Rsa,
+        rustls::ServerConfig::builder_with_provider(
+            CryptoProvider {
+                cipher_suites: vec![
+                    ffdhe::TLS_DHE_RSA_WITH_AES_128_GCM_SHA256,
+                    provider::cipher_suite::TLS_ECDHE_RSA_WITH_AES_128_GCM_SHA256,
+                ],
+                kx_groups: vec![&ffdhe::FFDHE2048_KX_GROUP, provider::kx_group::SECP256R1],
+                ..provider::default_provider()
+            }
+            .into(),
+        )
+        .with_safe_default_protocol_versions()
+        .unwrap(),
+    );
+
+    let (mut client, mut server) = make_pair_for_configs(client_config, server_config);
+    transfer(&mut client, &mut server);
+    assert!(server.process_new_packets().is_ok());
+    assert_eq!(
+        server
+            .negotiated_cipher_suite()
+            .unwrap()
+            .suite(),
+        CipherSuite::TLS_ECDHE_RSA_WITH_AES_128_GCM_SHA256
+    );
```

Comment:
Corresponding to my comment above, another server config with only `ffdhe::FFDHE2048_KX_GROUP` could be added to test to expect a fatal TLS alert of type insufficient_security at client side.